File: Grind75/13-ImplQueueUsingStacks/cpp/ImplQueueUsingStacks.cpp
#include "ImplQueueUsingStacks.h"
#include <stack>

MyQueue::MyQueue() {
}

void MyQueue::push(int x) {
  psh_stk.push(x);
}

int MyQueue::pop() {
  if (pop_stk.empty())
    flush_psh();
  int popped = pop_stk.top();
  pop_stk.pop();
  return popped;
}

int MyQueue::peek() {
  if (pop_stk.empty())
    flush_psh();
  return pop_stk.top();
}

bool MyQueue::empty() {
  return psh_stk.empty() && pop_stk.empty();
}

void MyQueue::flush_psh() {
  while (!psh_stk.empty()) {
    int top_psh = psh_stk.top();
    psh_stk.pop();
    pop_stk.push(top_psh);
  }
}

File: Grind75/13-ImplQueueUsingStacks/python/ImplQueueUsingStacks.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')

class MyQueue(object):
  enqueue: list 
  dequeue: list

  def __init__(self):
    self.enqueue = []
    self.dequeue = []

  def push(self, x: int) -> None:
    self.enqueue.append(x)

  def pop(self) -> int:
    if len(self.dequeue) == 0:
      while len(self.enqueue) > 0:
        self.dequeue.append(self.enqueue.pop())
    return self.dequeue.pop()

  def peek(self) -> int:
    if len(self.dequeue) == 0:
      while len(self.enqueue) > 0:
        self.dequeue.append(self.enqueue.pop())
    return self.dequeue[-1]

  def empty(self) -> bool:
    return len(self.enqueue) == 0 and len(self.dequeue) == 0


if __name__ == '__main__':
  soln = Solution()
  testCases = []
  
  for test in testCases:
    pass
  #print('TESTS PASSED!')

